block/sync.go

```diff
@@ -0,0 +1,121 @@
+package block
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/dymensionxyz/dymint/node/events"
+	"github.com/dymensionxyz/dymint/settlement"
+	uevent "github.com/dymensionxyz/dymint/utils/event"
+	"github.com/dymensionxyz/gerr-cosmos/gerrc"
+	"github.com/tendermint/tendermint/libs/pubsub"
+)
+
+// onNewStateUpdate will update the last submitted height and will update sequencers list from SL. After, it triggers syncing or validation, depending whether it needs to sync first or only validate.
+func (m *Manager) onNewStateUpdate(event pubsub.Message) {
+	eventData, ok := event.Data().(*settlement.EventDataNewBatch)
+	if !ok {
+		m.logger.Error("onReceivedBatch", "err", "wrong event data received")
+		return
+	}
+
+	// Update heights based on state update end height
+	m.LastSubmittedHeight.Store(eventData.EndHeight)
+	m.UpdateTargetHeight(eventData.EndHeight)
+
+	m.logger.Error("syncing")
+
+	// Update sequencers list from SL
+	err := m.UpdateSequencerSetFromSL()
+	if err != nil {
+		m.logger.Error("update bonded sequencer set", "error", err)
+	}
+
+	if eventData.EndHeight > m.State.Height() {
+		// Trigger syncing from DA.
+		m.triggerStateUpdateSyncing()
+	} else {
+		// trigger state update validation (in case no state update is applied)
+		m.triggerStateUpdateValidation()
+	}
+}
+
+// SyncLoop listens for syncing events (from new state update or from initial syncing) and syncs to the last submitted height.
+// It sends signal to validation loop for each synced state updated
+func (m *Manager) SyncLoop(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+
+		case <-m.syncingC:
+
+			m.logger.Info("syncing to target height", "targetHeight", m.LastSubmittedHeight.Load())
+
+			for currH := m.State.NextHeight(); currH <= m.LastSubmittedHeight.Load(); currH = m.State.NextHeight() {
+				// if we have the block locally, we don't need to fetch it from the DA
+				err := m.applyLocalBlock(currH)
+				if err == nil {
+					m.logger.Info("Synced from local", "store height", currH, "target height", m.LastSubmittedHeight.Load())
+					continue
+				}
+				if !errors.Is(err, gerrc.ErrNotFound) {
+					m.logger.Error("Apply local block", "err", err)
+				}
+
+				err = m.syncFromDABatch()
+				if err != nil {
+					m.logger.Error("process next DA batch", "err", err)
+				}
+
+				// if height havent been updated, we are stuck
+				if m.State.NextHeight() == currH {
+					return fmt.Errorf("stuck at height %d", currH)
+				}
+
+				m.logger.Info("Synced from DA", "store height", m.State.Height(), "target height", m.LastSubmittedHeight.Load())
+
+				// trigger state update validation, after each state update is applied
+				m.triggerStateUpdateValidation()
+
+				err = m.attemptApplyCachedBlocks()
```

Comment:
because this will try to apply upcoming blocks from cache if there are. i just moved it inside the loop, because otherwise it may not try to apply cached packets after every batch and it will only do it when finished syncing from da.

Reply:
but it seems like `AttemptApplyCachedBlocks` is anyway being called after `SyncFromDABatch` no? so anyway it will only attempt to apply cached blocks after that? also what I'm missing is:
1. assuming upcoming blocks are not from settlement, aren't those blocks should be applied in the p2p loop (i.e why we need to call it from here explicitly?) ?
2. why would the `syncFromSettlement` loop invoke a general `attemptApplyCacheBlocks` vs specifically calling it when we know there are new blocks (like we do with the `syncFromDABatch`)

also seems like blocks coming from settlement are not being feeded into the cache, and blocks that come from p2p are feeded into the cache which creates a weird flow.

IMO to make things clearer we must make sure all blocks follow the same funnel (i.e all blocks goes to cache, there are some way to trigger applying from cache by a callsite etc)

not sure about exact impl but it feels like the impl now is not uniform for all blocks appliance.

Reply:
we need to call AttemptApplyCachedBlocks from the settlement sync loop, because we need to check whether the next blocks are already in the cache (so we dont need to fetch them from da), and we do not want to wait for a next p2p event to apply them if already in cache. 
its true da blocks could be feeded into the same cache to unify block application, but this will be part of another pr (issue on that #1165)